pilot/pkg/serviceregistry/kube/controller/endpointslice.go

```diff
@@ -0,0 +1,292 @@
+	// in endpointsForPodInDifferentNS will thus be rejected by Pilot).
+	return append(endpointsForPodInSameNS, endpointsForPodInDifferentNS...)
+}
+
+func (e *endpointSliceController) InstancesByPort(c *Controller, svc *model.Service, reqSvcPort int,
+	labelsList labels.Collection) ([]*model.ServiceInstance, error) {
+	esLabelSelector := klabels.Set(map[string]string{discoveryv1alpha1.LabelServiceName: svc.Attributes.Name}).AsSelectorPreValidated()
+	var slices []*discoveryv1alpha1.EndpointSlice
+	err := cache.ListAllByNamespace(e.cache.informer.GetIndexer(), svc.Attributes.Namespace, esLabelSelector, func(i interface{}) {
+		slices = append(slices, i.(*discoveryv1alpha1.EndpointSlice))
+	})
+	if err != nil {
+		log.Infof("get endpoints(%s, %s) => error %v", svc.Attributes.Name, svc.Attributes.Namespace, err)
+		return nil, nil
+	}
+	if len(slices) == 0 {
+		return nil, nil
+	}
+
+	// Locate all ports in the actual service
+	svcPortEntry, exists := svc.Ports.GetByPort(reqSvcPort)
+	if !exists {
+		return nil, nil
+	}
+
+	var out []*model.ServiceInstance
+	for _, slice := range slices {
+		for _, e := range slice.Endpoints {
+			for _, a := range e.Addresses {
+				var podLabels labels.Instance
+				pod := c.pods.getPodByIP(a)
+				if pod != nil {
+					podLabels = configKube.ConvertLabels(pod.ObjectMeta)
+				}
+
+				// check that one of the input labels is a subset of the labels
+				if !labelsList.HasSubsetOf(podLabels) {
+					continue
+				}
+
+				az, sa, uid := "", "", ""
+				if pod != nil {
+					az = c.GetPodLocality(pod)
+					sa = kube.SecureNamingSAN(pod)
+					uid = fmt.Sprintf("kubernetes://%s.%s", pod.Name, pod.Namespace)
+				}
+				tlsMode := kube.PodTLSMode(pod)
+
+				// identify the port by name. K8S EndpointPort uses the service port name
+				for _, port := range slice.Ports {
+					var portNum uint32
+					if port.Port != nil {
+						portNum = uint32(*port.Port)
+					}
+
+					if port.Name == nil ||
+						svcPortEntry.Name == *port.Name {
+
+						out = append(out, &model.ServiceInstance{
+							Endpoint: model.NetworkEndpoint{
+								Address:     a,
+								Port:        int(portNum),
+								ServicePort: svcPortEntry,
+								UID:         uid,
+								Network:     c.endpointNetwork(a),
+								Locality:    az,
+							},
+							Service:        svc,
+							Labels:         podLabels,
+							ServiceAccount: sa,
+							TLSMode:        tlsMode,
+						})
+					}
+				}
+			}
+		}
+	}
+	return out, nil
+}
+
+type endpointSliceCache struct {
+	mu                         sync.RWMutex
+	endpointsByServiceAndSlice map[host.Name]map[string][]*model.IstioEndpoint
+}
+
+func newendpointSliceCache() *endpointSliceCache {
+	out := &endpointSliceCache{
+		endpointsByServiceAndSlice: make(map[host.Name]map[string][]*model.IstioEndpoint),
+	}
+	return out
+}
+
+func (e *endpointSliceCache) Update(hostname host.Name, slice string, endpoints []*model.IstioEndpoint) {
```

Comment:
QUestion: how do you delete the key?

Reply:
Good point. Right now its never deleted, just set to empty list. I think we can just delete if `endpoints` is empty